### PR TITLE
Add distributed voucher system to libremesh

### PR DESCRIPTION
**Pull request by gmarcos87**
Me and Luandro have been working in Pirania and have built this plugin for shared-state that allows vouchers db to be shared between nodes.

### CODE REVIEW

File: packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+--! along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+local JSON = require("luci.jsonc")
+
+local inputTable = JSON.parse(io.stdin:read("*all"))
+local header ='key,voucher,expiretime,uploadlimit,downloadlimit,amountofmacsallowed,usedmacs,\n'
+
+local outputLines = ''
+
+function macs2string(macs)
```
Comment (G10h4ck) [inline packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers]: have you considered `table.concat` ?

File: packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        if  key ~= #macs then
+            result = result..'+'
+        end
+    end
+    return result
+end
+
+
+for key,sharedData in pairs(inputTable) do
+    vo = sharedData.data
```
Comment (G10h4ck) [inline packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers]: Isn't `vo` a local variable?

File: packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+--!
+--! You should have received a copy of the GNU Affero General Public License
+--! along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+local fs = require("nixio.fs")
+local JSON = require("luci.jsonc")
+
+
+local function shell(command)
+    -- TODO(nicoechaniz): sanitize or evaluate if this is a security risk
```
Comment (G10h4ck) [inline packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers]: This comment seems to not apply here, did you copied this function from another place?

File: packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers (hunk 1 above)
Comment (gmarcos87) [reply packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers]: changed

File: packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers (hunk 2 above)
Comment (gmarcos87) [reply packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers]: changed

Comment (gmarcos87) [reply packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers]: changed

File: packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers (hunk 3 above)
Comment (gmarcos87) [reply packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers]: I use this function in many places, it was a function that Nico originally wrote and that for future references and replacement I copy as is. I remove the comment anyway.

File: packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+local outputTable ={"key,voucher,expiretime,uploadlimit,downloadlimit,amountofmacsallowed,usedmacs,"}
+
+local function shell(command)
+    local handle = io.popen(command)
+    local result = handle:read("*a")
+    handle:close()
+    return result
+end
+
+function macs2string(macs)
```
Comment (G10h4ck) [inline packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers]: I meant that this whole function could be substituted by table.concat, It's me or the function is still there?

File: packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+end
+
+local vouchers = JSON.parse(shell('voucher list_vouchers')) or {}
+local result = {}
+
+for _, voucher in pairs(vouchers) do
+    result[voucher.voucher] = voucher
+end
+
+io.popen("shared-state insert pirania", "w"):write(JSON.stringify(result))
```
Comment (G10h4ck) [inline packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers]: Add a new line at the end of the file please

File: packages/shared-state-pirania/files/etc/pirania/hooks/start/01-publish_and_sync [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+/etc/shared-state/publishers/shared-state-publish_vouchers
+shared-state sync pirania
```
Comment (G10h4ck) [inline packages/shared-state-pirania/files/etc/pirania/hooks/start/01-publish_and_sync]: new line at the end of the file is missin

**Comment by G10h4ck** [comment]
@nicopace @aparcar do you read the code before appriving pull requests or you approve them on faith that my review is enough?

**Comment by nicopace** [comment]
hehehe... i did the proper review... but good that u r around to catch the things you mentioned.

**Comment by aparcar** [comment]
I check if the PR possibly breaks other stuff, if not and the package doesn't concern "core" functions of LiMe I'm approving it

File: packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers (hunk 4 above)
Comment (gmarcos87) [reply packages/shared-state-pirania/files/etc/shared-state/hooks/pirania/generate_vouchers]: done in 3fd69c9c4af326ac64da7e012748fb60a35c68a1

File: packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers (hunk 5 above)
Comment (gmarcos87) [reply packages/shared-state-pirania/files/etc/shared-state/publishers/shared-state-publish_vouchers]: done in 9c23f93ecfabab041139f534d3d731e88b9398b9

File: packages/shared-state-pirania/files/etc/pirania/hooks/start/01-publish_and_sync (hunk 6 above)
Comment (gmarcos87) [reply packages/shared-state-pirania/files/etc/pirania/hooks/start/01-publish_and_sync]: done in 9c23f93ecfabab041139f534d3d731e88b9398b9

**Pull request merged**